### PR TITLE
[NOD-909] Add tests for double spends

### PR DESCRIPTION
**Pull request by someone235**

### CODE REVIEW

File: mempool/mempool_test.go [hunk 1]
```diff
@@ -795,6 +795,67 @@ func TestDoubleSpends(t *testing.T) {
...
+	defer teardownFunc()
+	harness := tc.harness
+
+	tx, err := harness.createTx(spendableOuts[0], uint64(txRelayFeeForTest), 1)
+	if err != nil {
+		t.Fatalf("unable to create transaction: %v", err)
+	}
+
+	dag := harness.txPool.cfg.DAG
+	block, err := mining.PrepareBlockForTest(dag, harness.dagParams, dag.TipHashes(), []*wire.MsgTx{tx.MsgTx()}, false)
```
Comment (stasatdaglabs) [inline mempool/mempool_test.go:812]: Please use prepareAndProcessBlock here.

File: blockdag/dag_test.go [hunk 2]
```diff
@@ -1117,3 +1117,161 @@ func TestIsDAGCurrentMaxDiff(t *testing.T) {
...
+	}
+	if isDelayed {
+		t.Fatalf("ProcessBlock: block2 " +
+			"is too far in the future")
+	}
+	if isOrphan {
+		t.Fatalf("ProcessBlock: block2 got unexpectedly orphaned")
+	}
+
+	// Check that a block will be rejected if it has a transaction from its past.
```
Comment (svarogg) [inline blockdag/dag_test.go:1188]: "from it's past" not very clear. I'd revise to "already existing in it's past".

File: blockdag/dag_test.go [hunk 3]
```diff
@@ -1117,3 +1117,161 @@ func TestIsDAGCurrentMaxDiff(t *testing.T) {
...
+	block3, err := PrepareBlockForTest(dag, []*daghash.Hash{block2.BlockHash()}, nil)
```
Comment (svarogg) [inline blockdag/dag_test.go:1189]: Give block1,2,3,3A names with meanings

File: mempool/mempool_test.go [hunk 4]
```diff
@@ -795,6 +795,67 @@ func TestDoubleSpends(t *testing.T) {
...
+	}
+	if isDelayed {
+		t.Fatalf("ProcessBlock: block " +
+			"is too far in the future")
+	}
+	if isOrphan {
+		t.Fatalf("ProcessBlock: block got unexpectedly orphaned")
+	}
+
+	// Check that a transaction that some of its outputs exists in the DAG UTXO
```
Comment (svarogg) [inline mempool/mempool_test.go:828]: I'm not sure I understand what "some of its outputs exists in the DAG UTXO" means

File: mempool/mempool_test.go [hunk 5]
```diff
@@ -809,24 +809,26 @@ func TestDoubleSpendsFromDAG(t *testing.T) {
...
-	}
-	if isOrphan {
-		t.Fatalf("ProcessBlock: block got unexpectedly orphaned")
+		t.Fatalf("ProcessTransaction: %s", err)
 	}
+	testPoolMembership(tc, doubleSpendTx, true, false, false)
 
-	// Check that a transaction that some of its outputs exists in the DAG UTXO
-	// is rejected from the mempool.
+	// If you send a transaction that some of its outputs exist in the DAG UTXO,
```
Comment (svarogg) [inline mempool/mempool_test.go:826]: UTXO -> UTXO-set.

**Pull request merged**